recipes/binutils/all/conanfile.py

```diff
@@ -1,70 +1,160 @@
+from conans import ConanFile, AutoToolsBuildEnvironment, tools
+from conans.errors import ConanInvalidConfiguration
+import functools
 import os
-import re
-from glob import glob
 
-from conan import ConanFile
-from conan.tools.gnu import Autotools, AutotoolsDeps, AutotoolsToolchain
-from conans import tools
-from conans.errors import ConanInvalidConfiguration
 
 required_conan_version = ">=1.43.0"
 
 
 class BinutilsConan(ConanFile):
     name = "binutils"
-    description = "a collection of binary tools"
-    topics = ("gnu", "compilation", "tools")
-    url = "https://github.com/conan-io/conan-center-index"
-    homepage = "https://www.gnu.org/software/binutils"
+    description = "The GNU Binutils are a collection of binary tools."
     license = "GPL-2.0-or-later"
-
+    url = "https://github.com/conan-io/conan-center-index/"
+    homepage = "https://www.gnu.org/software/binutils"
+    topics = ("binutils", "ld", "linker", "as", "assembler", "objcopy", "objdump")
     settings = "os", "arch", "compiler", "build_type"
 
-    _source_subfolder = "source_subfolder"
+    options = {
+        "multilib": [True, False],
+        "with_libquadmath": [True, False],
+        "target_arch": "ANY",
+        "target_os": "ANY"
+    }
+
+    default_options = {
+        "multilib": True,
+        "with_libquadmath": True,
+        "target_arch": None,  # Initialized in config_options
+        "target_os": None,  # Initialized in config_options
+    }
+
+    @property
+    def _source_subfolder(self):
+        return "source_subfolder"
+
+    @property
+    def _settings_build(self):
+        return getattr(self, "settings_build", self.settings)
+
+    @property
+    def _settings_target(self):
+        return getattr(self, "settings_target", None) or self.settings
```

Comment:
I don't think this is going to work... but I don't really know how `binutils` works. So, please, read my lines in the context of a cross-compiler (Android NDK, emsdk,...) and, if this is not the case for `binutils`, just ignore all the following.

(After reading previous paragraph, continue at your own risk 😄 )

Here (**while creating the package**) we are assigning to `option.target_os` the value of the _host_ os, which means that it is targeting the same OS where it runs. We built the package for the _host_ OS and the value of `option.target_os` is that same OS. This is essentially wrong is the purpose is to target a different OS from the one where `binutils` is running (~crosscompiling scenario).

When I suggest using an option it is because the binary (the files in the Conan package) is different depending on the target system, like containing a different sysroot... libraries pre-compiled for that target system. With this approach, at the time of generating the package the creator needs to provide a value explicitly:

 * **`pr-linux-to-windows`**: profile building Linux package that targets Windows
   ```ini
   [settings]
   os=Linux
   [options]
   binutils:target_os=Windows
   ```
   This would generate a `binutils` package that runs in Linux and contains all the DLLs required to target Windows:
   ```
   conan create binutils/xxxx --profile:host=pr-linux-to-windows
   ```

 * **`pr-linux-to-macos`**: profile building Linux package that targets Windows
   ```ini
   [settings]
   os=Linux
   [options]
   binutils:target_os=macos
   ```
   This would generate a `binutils` package that runs in Linux and contains all the DYLIBs required to target Macos:
   ```
   conan create binutils/xxxx --profile:host=pr-linux-to-macos
   ```
   
Both packages are different because they have different artifacts inside. As the settings are equal (both are to be run in Linux) we can only have different packageIDs using an option. We use that option (explicitly set at the time of building) to choose the artifacts to include in the resulting package.

---

Then we move to the consumption stage. In order for Conan to select the proper package, the consumer needs to provide settings and options that compute the desired packageID. As this is used as a build-requires in the build context, we can use the same profiles we used to generate those packages: 
 * `--profile:build=pr-linux-to-windows` will select the package that runs in Linux and is capable of targeting Windows
 * `--profile:build=pr-linux-to-macos` will select the package that runs in Linux and is capable of targeting Windows

And the consumer needs to provide the profile for the _host_ context as well:
 * if targeting windows, the _host_ profile should match Windows: `--profile:host=windows`
 * if targeting macos, the _host_ profile should match Macos: `--profile:host=macos`

**All of this is implicit knowledge on the consumer side**, the target architecture at consumption time used by `binutils` in the _build_ context needs to match the _host_ context. **Our recipe can only inform/raise** the user about a mismatch, but it cannot modify settings (it would be chicken-egg problem):

```python
def validate(self):
    if hasattr(self, 'settings_target'):  # We can only check when USED as build-requires
       if self.options.target_os != self.settings_target.os:
          raise ConanInvalid.... "Mismatch between target architecture and the one you are building your packages for"
``` 

Please, forgive me if `binutils` is not something like a cross-compiler.

Reply:
`binutils` is a cross-compiler, so build/host/target can be all different from each other

In this recipe, I basically do:
```python
class BinutilsConan(ConanFile):
    @property
    def _settings_target(self):
        return getattr(self, "settings_target", None) or self.settings
        
    def config_options(self):
        self.options.target_arch = str(self._settings_target.arch)
        self.options.target_os = str(self._settings_target.os)
        
    def validate(self):
        if self.options.target_arch not in self.settings.arch.values_range:
            raise ConanInvalidConfiguration(f"target_arch={self.options.target_arch} is invalid (possibilities={self.settings.arch.values_range})")
        if self.options.target_os not in self.settings.os.values_range:
            raise ConanInvalidConfiguration(f"target_os={self.options.target_os} is invalid (possibilities={self.settings.os.values_range})")
```

By assigning `config_option.target_{arch,os}` in `config_options` (overridable by the user), isn't conan assigning different package ids?

I created the following table:


| build    |   host    |   target    | package id                                 | command
| -------- | --------- | ----------- | ------------------------------------------ | ------------
| Linux    | Linux     | Linux       | `d0c627b96beba6c4f548b2e66247eece9701544a` | `conan create . binutils/2.38@`
| Linux    | Linux     | Linux       | `d0c627b96beba6c4f548b2e66247eece9701544a` | `conan create . binutils/2.38@ -o target_os=Linux`
| Linux    | Linux     | Windows     | `ac7cb2c348c6917e8ee2a45dd0268fedfc31d6d1` | `conan create . binutils/2.38@ -o target_os=Windows`
| Linux    | Windows   | Linux       | `8140b7dcc9c74a04fdcff4f118c8cc3b5dcbd5b4` | `conan create . binutils/2.38@ -pr:h mingw64 -pr:b default64 -o binutils:target_os=Linux`
| Linux    | Windows   | FreeBSD     | `5fbdaa86696ff852b8b301d8c15039a082ea75a5` | `conan create . binutils/2.38@ -pr:h mingw64 -pr:b default64 -o binutils:target_os=FreeBSD`

So all different host/target combinations result in different package ids, isn't that what we want?

If I'm wrong somewhere, can you please give me an explicit example?